ai_services/anomaly_detection/data_preprocessing_examples/oci_data_flow_based_examples/example_code/df_driver.py

```diff
@@ -0,0 +1,196 @@
+from example_code.one_hot_encoding import one_hot_encoding
+from example_code.string_transformations import string_transformation
+from example_code.format_timestamp import format_timestamp
+from example_code.temporal_differencing import temporal_differencing
+from example_code.normalization import normalize_data
+from example_code.fixed_window_batching import windowing
+from example_code.sliding_window_aggregation import aggregation
+from example_code.one_hot_encoding import one_hot_encoding
+from example_code.pivoting import spark_pivoting
+from example_code.sharding import sharding
+
+from example_code.time_series_merge import time_series_merge
+from example_code.time_series_join import time_series_join
+
+signer = oci.auth.signers.get_resource_principals_signer()
+data_flow_client = oci.data_flow.DataFlowClient(config={}, signer=signer)
+object_storage_client = \
+    oci.object_storage.ObjectStorageClient(config={}, signer=signer)
+SINGLE_DATAFRAME_PROCESSING = "singleDataFrameProcessing"
+COMBINE_DATAFRAMES = "combineDataFrames"
+JOIN = "join"
+MERGE = "merge"
+RESERVED_METADATA = ['distinct_categories']
+
+
+def get_token_path(spark):
+    token_key = "spark.hadoop.fs.oci.client.auth.delegationTokenPath"
+    token_path = spark.sparkContext.getConf().get(token_key)
+    return token_path
+
+
+def get_authenticated_client(token_path, client):
+    if token_path is None:
+        # You are running locally, so use our API Key.
+        config = oci.config.from_file()
+        authenticated_client = client(config)
+    else:
+        # You are running in Data Flow, so use our Delegation Token.
+        with open(token_path) as fd:
+            delegation_token = fd.read()
+        signer = oci.auth.signers.InstancePrincipalsDelegationTokenSigner(
+            delegation_token=delegation_token
+        )
+        authenticated_client = client(config={}, signer=signer)
+    return authenticated_client
+
+
+def get_object(object_storage_client, namespace, bucket, file):
+    get_resp = object_storage_client.get_object(namespace, bucket, file)
+    return get_resp.data.text
+
+
+def parse_and_process_data_preprocessing_config(object_storage_client, spark, get_resp, output_path):
+    dfs = dict()
+    try:
+        contents = json.loads(get_resp)
+        '''
+        Deal with input sources
+        '''
+        input_sources = contents["inputSources"]
+
+        # build dictonary: df name: df
+        for source in input_sources:
+            if source["type"] == "object-storage":
+                raw_data = get_object(object_storage_client, source["namespace"], source["bucket"], source["objectName"])
+                data = StringIO(raw_data)
+                pd_df = pd.read_csv(data, sep=",")
+                df = spark.createDataFrame(pd_df)
+                df = df.select([F.col(x).alias(x.lower()) for x in df.columns])
+                dfs[source["dataframeName"]] = df
+            elif source["type"] == "database":
```

Comment:
I think the format of `oracle` when reading the data ([ref](https://github.com/bug-catcher/oci-data-science-ai-samples/blob/bfdcc336b0ad76601db9daa225b99bcbd256a3be/ai_services/anomaly_detection/data_preprocessing_examples/oci_data_flow_based_examples/example_code/df_driver.py#L89)) says that this is going to be Oracle type database. So, please see if you could change this to `oracle` instead. 

[Minor] [line 90](https://github.com/bug-catcher/oci-data-science-ai-samples/blob/bfdcc336b0ad76601db9daa225b99bcbd256a3be/ai_services/anomaly_detection/data_preprocessing_examples/oci_data_flow_based_examples/example_code/df_driver.py#L90) can be replaced with the following
```suggestion
            elif source["type"] == "database":
              source.pop("type")
              df = spark.read \
                      .format("oracle") \
                      .options(**source) \
                      .load()
```

Reply:
> I think the format of `oracle` when reading the data ([ref](https://github.com/bug-catcher/oci-data-science-ai-samples/blob/bfdcc336b0ad76601db9daa225b99bcbd256a3be/ai_services/anomaly_detection/data_preprocessing_examples/oci_data_flow_based_examples/example_code/df_driver.py#L89)) says that this is going to be Oracle type database. So, please see if you could change this to `oracle` instead.
> 
> [Minor] [line 90](https://github.com/bug-catcher/oci-data-science-ai-samples/blob/bfdcc336b0ad76601db9daa225b99bcbd256a3be/ai_services/anomaly_detection/data_preprocessing_examples/oci_data_flow_based_examples/example_code/df_driver.py#L90) can be replaced with the following

Sorry I didn't get what do you mean by `change this to oracle`, since line 84 it's already `oracle`? Also, what does the `source.pop("type")` do?

Reply:
By discussing with Ravi, we clarified that we can change the "type" from "database" into "oracle" if customers are using ATP or ADW, because maybe in the future we want to support other sources like AWS S3 and GCP cloud storage. I will update the doc accordingly